tools/workspace-plugin/src/executors/verify-packaging/executor.ts

```diff
@@ -1,11 +1,167 @@
-import { VerifyPackagingExecutorSchema } from './schema';
-
-const runExecutor: PromiseExecutor<VerifyPackagingExecutorSchema> = async options => {
-  console.log('Executor ran for VerifyPackaging', options);
-  return {
-    success: true,
-  };
+import { type ExecutorContext, type PromiseExecutor, logger, serializeJson } from '@nx/devkit';
+import { spawnSync } from 'node:child_process';
+
+import micromatch from 'micromatch';
+
+import { type VerifyPackagingExecutorSchema } from './schema';
+import { join } from 'node:path';
+
+const runExecutor: PromiseExecutor<VerifyPackagingExecutorSchema> = async (schema, context) => {
+  const options = normalizeOptions(schema, context);
+
+  const success = await runVerifyPackaging(options, context);
+
+  return { success };
 };
 
 export default runExecutor;
+
+// ========
+
+interface NormalizedOptions extends ReturnType<typeof normalizeOptions> {}
+
+async function runVerifyPackaging(options: NormalizedOptions, context: ExecutorContext): Promise<boolean> {
+  const tags = new Set(options.project.tags ?? []);
+  const isPublic = tags.has('npm:public');
+
+  // no need to check if package is not being published yet
+  if (!isPublic) {
+    return true;
+  }
+
+  const packOutput = npmPackOutput(options, context);
+
+  const issues = assertions(packOutput, options, tags);
+
+  if (issues.length === 0) {
+    return true;
+  }
+
+  logger.error(`Package verification failed!`);
+  logger.error(serializeJson(issues));
+
+  return false;
+}
+
+function normalizeOptions(schema: VerifyPackagingExecutorSchema, context: ExecutorContext) {
+  const defaults = {};
+  const project = context.projectsConfigurations!.projects[context.projectName!];
+  const isProduction = Boolean(process.env.FLUENT_PROD_BUILD);
+
+  /**
+   * @see https://docs.npmjs.com/cli/v10/commands/npm-publish#files-included-in-package
+   */
+  const alwaysPublishedFiles = ['LICENSE', 'package.json', 'README.md'];
+  const rootConfigFiles = [
+    'just.config.[jt]s',
+    'jest.config.[jt]s',
+    '.eslintrc.(js|json)',
+    'project.json',
+    '.babelrc.json',
+    '.swcrc',
+    'tsconfig(.*)?.json',
+  ];
+  const nonProdAssets = ['assets/', 'docs/*', 'temp/*', 'bundle-size/*', '.storybook/*', 'stories/*'];
+
+  const filePatterns = { alwaysPublishedFiles, rootConfigFiles, nonProdAssets };
+
+  return { ...defaults, ...schema, project, isProduction, filePatterns };
+}
+
+function npmPackOutput(options: NormalizedOptions, context: ExecutorContext) {
+  const npmPackResult = spawnSync('npm', ['pack', '--dry-run'], { cwd: join(context.root, options.project.root) });
+
+  const processedResult = npmPackResult.output
+    .toString()
+    .replace(/\bnpm notice\b\s+[\d.]+[MkB]+\s+/gi, '')
+    .replace(/[ ]+/g, '');
+  return processedResult.split('\n');
+}
+
+function assertions(
+  npmPackResult: string[],
+  options: NormalizedOptions,
+  tags: Set<string>,
+): Array<{ matches: string[]; message: string }> {
```

Comment:
Is the use of both array literal syntax `[]` and `Array<...>` in the same line intentional? Could we standardize on one notation?

Reply:
> Is the convention to use the literal syntax [] for primitive types and Array<...>

yes - no lint rules in place to enforce in place. we should establish these rules but at this stage it's ultra low priority. the style of array declaration doesn't affect anything ( especially in tooling packages )